Avoid scatter for single-device case in DDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46304 Avoid scatter for single-device case in DDP**

In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

rohan-varma committed:
Avoid scatter for single-device case in DDP

In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3781,6 +3781,56 @@ def forward(self, x):
+                        return _self.lin(x.t)
+                    else:
+                        self.assertTrue(len(x), expected_len)
+                        self.assertTrue(x[0].device == x[1].device)
```
Shouldn't we pass in the expected device to the constructor of ToyModel and validate it is correct here?

rohan-varma replied:
We could do this, but the test is basically expected to validate that the device is the current rank of the process, as we are testing single GPU per process. The next line asserts that the input is on the expected device.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3781,6 +3781,56 @@ def forward(self, x):
+            model(inp, tuple)
+            # List CPU input, should be moved to proper device before call to
+            # forward.
+            inp = [torch.randn(10, 10) for _ in range(expected_len)]
+            model(inp, list)
```
Can we add tests for dict and namedtuple as well?

rohan-varma replied:
Added these in the latest version of the diff.

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -660,10 +660,11 @@ def forward(self, *inputs, **kwargs):
-            inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
             if len(self.device_ids) == 1:
+                inputs, kwargs = self.to_kwargs(inputs, kwargs, self.device_ids[0])
                 output = self.module(*inputs[0], **kwargs[0])
```
Curious, any reason we still need to return `inputs` and `kwargs` as a list?

rohan-varma replied:
I was mostly doing that to keep parity with the current version, but we can probably remove it in this code path.

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Avoid scatter for single-device case in DDP"


In the case that a single process operates only on one GPU, we can
avoid this scatter and instead replace it with a recursive version of `to`
which transfers the input tensors to the correct device.

The implementation of `_recursive_to` is modeled after `scatter` in https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/scatter_gather.py, in order to keep parity with the previous conventions (i.e. custom types not having their tensors moved).

Differential Revision: [D24296377](https://our.internmc.facebook.com/intern/diff/D24296377/)

[ghstack-poisoned]

ngimel commented:
@rohan-varma what's the motivation behind this PR? Does scatter for a single device incur some performance penalty? As #49819 says, previously it was possible to overlap h2d transfers with computation, now since the transfers happen on the default stream (`.to()` uses the default stream, as opposed to side stream in Scatter) this overlap is not possible.
Of course, current implementation can be fixed to use side stream, but I'm wondering if it's worth code complexity.
